### PR TITLE
feat(storage): Support for repairing the size of a split sst based on table stats

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Hummock corrects the commit_epoch commit sst based on the current compaction group mapping and places the newly generated sst in the correct compaction group by `split sst`.

Wrong commit sst size may cause compaction jitter in some compaction group, this PR re-estimates the size of the newly generated sst based on table_stats.

notes: After this pr, the group by logic in the shared buffer compact can be removed and CN only submit one grouped sst to the meta at each checkpoint.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/meta/src/hummock/manager/commit_epoch.rs [hunk 1]
```diff
@@ -413,8 +428,38 @@ impl HummockManager {
...
+                    .sum();
+
+                let mut branch_sst = split_sst(
+                    &mut sst.sst_info,
+                    &mut new_sst_id,
+                    origin_sst_size - new_sst_size,
+                    new_sst_size,
+                );
+
+                // FIXME(li0k): We would change table_ids inside the `split_sst` function
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/commit_epoch.rs:460]: +1

File: proto/hummock.proto [hunk 2]
```diff
@@ -913,6 +913,8 @@ message TableStats {
   int64 total_key_size = 1;
   int64 total_value_size = 2;
   int64 total_key_count = 3;
+
+  uint64 compressed_size_in_sstable = 4;
```
[the remark this answers is not in the document]
Comment (hzxa21) [reply proto/hummock.proto:917]: Also, can you add a comment to document that this field is only filled and used by CN flushes, not compactor compaction.

File: src/storage/src/hummock/sstable/builder.rs [hunk 3]
```diff
@@ -235,6 +230,20 @@ impl<W: SstableWriter, F: FilterBuilder> SstableBuilder<W, F> {
...
+            .block_metas
+            .iter()
+            .filter(|block_meta| block_meta.table_id().table_id() == self.last_table_id.unwrap())
+            .map(|block_meta| block_meta.len as u64)
+            .sum::<u64>();
+
+        if self.last_table_id.is_none() || self.last_table_id.unwrap() != table_id {
+            self.table_ids.insert(table_id);
+            self.finalize_last_table_stats();
+            self.last_table_id = Some(table_id);
```
[the remark this answers is not in the document]
Comment (Li0k) [reply src/storage/src/hummock/sstable/builder.rs:244]: I confirmed that in L199-L232 `last_table_id` is only used in the log and  no logic involved, I added table_id to the log.

The change is just to make sure that the table_id toggle is after the build_block.

File: src/meta/src/hummock/manager/commit_epoch.rs (hunk 1 above)
Comment (Li0k) [reply src/meta/src/hummock/manager/commit_epoch.rs]: refactor it , PTAL

File: src/meta/src/hummock/manager/compaction_group_manager.rs [hunk 4]
```diff
@@ -605,16 +605,6 @@ impl HummockManager {
         drop(compaction_guard);
         self.report_compact_tasks(canceled_tasks).await?;
 
-        // Don't trigger compactions if we enable deterministic compaction
```
Comment (Li0k) [inline src/meta/src/hummock/manager/compaction_group_manager.rs:608]: after https://github.com/risingwavelabs/risingwave/pull/18053/commits/2ccdb93809f65596b1aade9d42b48a9f900d0a16

SstableInfo's `table_ids` have already been modified in `split_sst`, so there is no need to trigger SpaceRecliam.

File: src/storage/src/hummock/sstable/builder.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hzxa21) [reply src/storage/src/hummock/sstable/builder.rs:238]: There is a `return` in L215. That means the table id will not be changed before returning. Will this be an issue?

File: src/meta/src/hummock/manager/compaction_group_manager.rs (hunk 4 above)
Comment (hzxa21) [reply src/meta/src/hummock/manager/compaction_group_manager.rs:608]: What if all the table ids of a SST is split out? I think in our current implementation, we will have two SST: one with empty table id and the other one with identical table id as before the split.

File: src/storage/src/hummock/sstable/builder.rs [hunk 5]
```diff
@@ -529,6 +530,35 @@ impl<W: SstableWriter, F: FilterBuilder> SstableBuilder<W, F> {
...
+                last_total_compressed_size += block_meta.len as u64;
+            }
+
+            if last_total_compressed_size != 0 {
+                self.table_stats
+                    .get_mut(&last_table_id.unwrap())
+                    .unwrap()
+                    .total_compressed_size = last_total_compressed_size;
+            }
+        }
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/builder.rs:560]: nits: there are so many `if`S in this simple logic, how about making it more concise:
```rust
if !meta.block_metas.is_empty() {
    // fill total_compressed_size

    let mut last_table_id = meta.block_metas[0].table_id();
    let mut last_table_stats = self.table_stats.get_mut(&last_table_id).unwrap();
    for block_meta in &meta.block_metas {
        let block_table_id = block_meta.table_id();
        if last_table_id != block_table_id {
            last_table_id = block_table_id;
            last_table_stats = self.table_stats.get_mut(&last_table_id).unwrap();
        }
        last_table_stats.total_compressed_size += block_meta.len as u64;
    }
}
```

File: src/meta/src/hummock/manager/compaction_group_manager.rs (hunk 4 above)
Comment (Li0k) [reply src/meta/src/hummock/manager/compaction_group_manager.rs:608]: SstableInfo with empty table_ids will be cleaned up immediately in the apply_version_delta phase.

**Review by zwang28 — APPROVED** [review]

**Comment by Li0k** [comment]
https://github.com/risingwavelabs/risingwave/issues/15973

**Pull request merged**